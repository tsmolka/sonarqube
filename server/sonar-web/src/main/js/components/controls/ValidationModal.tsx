/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { withFormik, Form, FormikActions, FormikProps } from 'formik';
import Modal from './Modal';
import { ResetButtonLink, SubmitButton } from '../../components/ui/buttons';
import DeferredSpinner from '../common/DeferredSpinner';
import { translate } from '../../helpers/l10n';

interface InnerFormProps<Values> {
  children: (props: FormikProps<Values>) => React.ReactNode;
  confirmButtonText: string;
  header: string;
  initialValues: Values;
}

interface Props<Values> extends InnerFormProps<Values> {
  isInitialValid?: boolean;
  onClose: () => void;
  validate: (data: Values) => void | object | Promise<object>;
  onSubmit: (data: Values) => void | Promise<void>;
}

export default class ValidationModal<Values> extends React.PureComponent<Props<Values>> {
  handleSubmit = (data: Values, { setSubmitting }: FormikActions<Values>) => {
    const result = this.props.onSubmit(data);
    if (result) {
      result.then(
        () => {
          setSubmitting(false);
          this.props.onClose();
        },
        () => {
          setSubmitting(false);
        }
      );
    } else {
      setSubmitting(false);
      this.props.onClose();
    }
  };

  render() {
    const { header } = this.props;

    const InnerForm = withFormik<InnerFormProps<Values>, Values>({
      handleSubmit: this.handleSubmit,
      isInitialValid: this.props.isInitialValid,
      mapPropsToValues: props => props.initialValues,
      validate: this.props.validate
    })(props => (
      <Form>
        <div className="modal-head">
          <h2>{props.header}</h2>
        </div>

        <div className="modal-body">{props.children(props)}</div>

        <footer className="modal-foot">
          <DeferredSpinner className="spacer-right" loading={props.isSubmitting} />
          <SubmitButton disabled={props.isSubmitting || !props.isValid || !props.dirty}>
            {props.confirmButtonText}
          </SubmitButton>
          <ResetButtonLink disabled={props.isSubmitting} onClick={this.props.onClose}>
            {translate('cancel')}
          </ResetButtonLink>
        </footer>
      </Form>
    ));

    return (
      <Modal contentLabel={header} onRequestClose={this.props.onClose}>
        <InnerForm
          confirmButtonText={this.props.confirmButtonText}
          header={header}
          initialValues={this.props.initialValues}>
          {this.props.children}
        </InnerForm>
      </Modal>
    );
  }
}
