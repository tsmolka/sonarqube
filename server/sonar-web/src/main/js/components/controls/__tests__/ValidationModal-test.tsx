/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import { FormikProps } from 'formik';
import ValidationModal from '../ValidationModal';

it('should render correctly', () => {
  const { wrapper, inner } = getWrapper();
  expect(wrapper).toMatchSnapshot();
  expect(inner).toMatchSnapshot();
});

interface Values {
  field: string;
}

function getWrapper(props = {}) {
  const wrapper = shallow(
    <ValidationModal
      confirmButtonText="confirm"
      header="title"
      initialValues={{ field: 'foo' }}
      isInitialValid={true}
      onClose={jest.fn()}
      validate={(values: Values) => ({ field: values.field.length < 2 && 'Too small' })}
      onSubmit={jest.fn(() => Promise.resolve())}
      {...props}>
      {(props: FormikProps<Values>) => (
        <form onSubmit={props.handleSubmit}>
          <input
            onChange={props.handleChange}
            onBlur={props.handleBlur}
            name="field"
            type="text"
            value={props.values.field}
          />
        </form>
      )}
    </ValidationModal>
  );
  return {
    wrapper,
    inner: wrapper
      .childAt(0)
      .dive()
      .dive()
      .dive()
  };
}
