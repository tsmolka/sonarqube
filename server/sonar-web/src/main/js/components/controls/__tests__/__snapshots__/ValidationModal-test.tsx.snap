// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<Modal
  contentLabel="title"
  onRequestClose={[MockFunction]}
>
  <C
    confirmButtonText="confirm"
    header="title"
    initialValues={
      Object {
        "field": "foo",
      }
    }
  />
</Modal>
`;

exports[`should render correctly 2`] = `
<Form>
  <div
    className="modal-head"
  >
    <h2>
      title
    </h2>
  </div>
  <div
    className="modal-body"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        name="field"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="foo"
      />
    </form>
  </div>
  <footer
    className="modal-foot"
  >
    <DeferredSpinner
      className="spacer-right"
      loading={false}
      timeout={100}
    />
    <SubmitButton
      disabled={true}
    >
      confirm
    </SubmitButton>
    <ResetButtonLink
      disabled={false}
      onClick={[MockFunction]}
    >
      cancel
    </ResetButtonLink>
  </footer>
</Form>
`;
